share/availability/discovery/discovery.go

```diff
@@ -245,11 +245,14 @@ func (d *Discovery) discover(ctx context.Context) bool {
 	var wg errgroup.Group
 	// limit to minimize chances of overreaching the limit
 	wg.SetLimit(int(d.set.Limit()))
-	defer wg.Wait() //nolint:errcheck
 
 	// stop discovery when we are done
 	findCtx, findCancel := context.WithCancel(ctx)
-	defer findCancel()
+	defer func() {
+		// some workers could still be running, wait them to finish before canceling findCtx
+		wg.Wait() //nolint:errcheck
+		findCancel()
```

Comment:
I thought one of the parts of the solution was cancelling connections that were taking too long - this happens in the real world mostly because of NAT we assume, and wont really show in tests

But since we are no longer "filling pending spots" with ongoing connection attempts, I assume this is no longer a concern?

Reply:
connections take different context, not the one being canceled. It is made this way to allow ongoing handshake to finish, since it could be already half done.